### PR TITLE
Fix move emulated classes copy assignment

### LINKED ISSUE #1434: cuSTL: memcopy from device buffer to device buffer fails
When a reference to a device buffer is assigned to another reference of a device buffer of the same type the following error message arises:
```
/home/burau/picongpu/src/picongpu/../libPMacc/include/cuSTL/container/DeviceBuffer.hpp(114): error: incomplete type is not allowed
          detected during:
            instantiation of "PMacc::container::DeviceBuffer<Type, T_dim> &PMacc::container::DeviceBuffer<Type, T_dim>::operator=(const HBuffer &) [with Type=picongpu::float_X, T_dim=3, HBuffer=boost::rv<PMacc::container::DeviceBuffer<picongpu::float_X, 3>>]" 
(63): here
            instantiation of "PMacc::container::DeviceBuffer<Type, T_dim> &PMacc::container::DeviceBuffer<Type, T_dim>::operator=(PMacc::container::DeviceBuffer<Type, T_dim> &) [with Type=picongpu::float_X, T_dim=3]" 
/home/burau/picongpu/src/picongpu/include/plugins/particleCalorimeter/ParticleCalorimeter.hpp(425): here
            instantiation of "void picongpu::ParticleCalorimeter<ParticlesType>::notify(uint32_t) [with ParticlesType=picongpu::PIC_Electrons]" 
/home/burau/picongpu/src/picongpu/include/plugins/particleCalorimeter/ParticleCalorimeter.hpp(419): here
            implicit generation of "picongpu::ParticleCalorimeter<ParticlesType>::~ParticleCalorimeter() [with ParticlesType=picongpu::PIC_Electrons]" 
/home/burau/picongpu/src/picongpu/include/plugins/particleCalorimeter/ParticleCalorimeter.hpp(419): here
            instantiation of class "picongpu::ParticleCalorimeter<ParticlesType> [with ParticlesType=picongpu::PIC_Electrons]" 
/home/burau/picongpu/src/picongpu/include/plugins/particleCalorimeter/ParticleCalorimeter.hpp(419): here
```

ParticleCalorimeter.hpp(425):
`*this->dBufCalorimeter = *this->dBufLeftParsCalorimeter;`
Definition:
```
typedef PMacc::container::DeviceBuffer<float_X, DIM3> DBufCalorimeter;

/* device calorimeter buffer for a single gpu */
DBufCalorimeter* dBufCalorimeter;
/* device calorimeter buffer for all particles which have left the simulation volume  */
DBufCalorimeter* dBufLeftParsCalorimeter;
```

This file can be found in my remote `feat-ParticleDetector` branch.
 @psychocoderHPC  @Flamefire

### PR DESCRIPTION
**Pull request by Flamefire**
This supersedes #1432 and closes #1434.

As the documentation of boost.Move states (http://www.boost.org/doc/libs/1_49_0/doc/html/move/how_the_library_works.html), one has to use `BOOST_COPY_ASSIGN_REF` for the copy assignment. This was not done in #1375 and #1401 and resulted in failures as the one in #1434/#1376

With this PR #1376 compiles on hypnos5

Note: Minor changes to have `HostBuffer.hpp` and `DeviceBuffer.hpp` symmetric for easier comparison

TLDR: RTFM :+1: 

@Heikman

### CODE REVIEW

**Comment by Heikman** [comment]
Very nice investigation! thanks a lot
